src/fast_array_utils/_plugins/numba_sparse.py

```diff
@@ -0,0 +1,237 @@
+# SPDX-License-Identifier: MPL-2.0
+# taken from https://github.com/numba/numba-scipy/blob/release0.4/numba_scipy/sparse.py
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, cast
+
+import numba.core.types as nbtypes
+import numpy as np
+from numba.core import cgutils
+from numba.core.imputils import impl_ret_borrowed
+from numba.extending import (
+    NativeValue,
+    box,
+    intrinsic,
+    make_attribute_wrapper,
+    models,
+    overload,
+    overload_attribute,
+    overload_method,
+    register_model,
+    typeof_impl,
+    unbox,
+)
+from scipy import sparse
+
+
+if TYPE_CHECKING:
+    from collections.abc import Callable
+    from typing import Any, ClassVar
+
+    from llvmlite.ir import IRBuilder, Value
+    from numba.core.base import BaseContext
+    from numba.core.extending import BoxContext, TypingContext, UnboxContext
+    from numba.core.typing.templates import Signature
+    from numba.core.typing.typeof import _TypeofContext
+    from numpy.typing import NDArray
+
+    from fast_array_utils.types import CSBase
+
+
+class CS2DType(nbtypes.Type):
+    """A Numba `Type` modeled after the base class `scipy.sparse.compressed._cs_matrix`."""
+
+    name: ClassVar[str]
+    cls: ClassVar[type[CSBase]]
+
+    @classmethod
+    def instance_class(
+        cls,
+        data: NDArray[np.number[Any]],
+        indices: NDArray[np.integer[Any]],
+        indptr: NDArray[np.integer[Any]],
+        shape: tuple[int, int],
+    ) -> CSBase:
+        return cls.cls((data, indices, indptr), shape, copy=False)
+
+    def __init__(self, dtype: nbtypes.Type) -> None:
+        self.dtype = nbtypes.DType(dtype)
+        self.data = nbtypes.Array(dtype, 1, "A")
+        self.indices = nbtypes.Array(nbtypes.int32, 1, "A")
+        self.indptr = nbtypes.Array(nbtypes.int32, 1, "A")
+        self.shape = nbtypes.UniTuple(nbtypes.int64, 2)
+        super().__init__(self.name)
```

Comment:
Same thing here about the `__init__` method - might be worth explaining the architecture of this class since it involves a bit of indirect handling to actually be used

Reply:
(And this paradigm is not covered in the example page)

Reply:
I explained things more in the docstring and added comments to all parts of this file.